File: LeetCode/Medium/0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int start = 0, end = nums.size()-1;

        while(start <= end){
            int mid = (start+end)/2;
            
            if(nums[mid]==target)
                return true;
            if(nums[mid]==nums[end]){
                end--;
                continue;
            }
            // left sorted
            if(nums[start] <= nums[mid]){
                if(nums[start]<=target && target<nums[mid])
                    end = mid - 1;
                else
                    start = mid + 1;
            }
            else{
                if(nums[mid]<target && target<=nums[end])
                    start = mid + 1;
                else
                    end = mid - 1;
            }
        }
        return false;
    }
};

File: LeetCode/Medium/0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        start, end = 0, len(nums) - 1

        while start <= end:
            mid = (start+end)//2
            print(mid)
            if nums[mid] == target:
                return True
            # fail to find sorted array
            if nums[mid] == nums[end]: 
                end -= 1
                continue
            # left sorted
            if nums[start] <= nums[mid]:
                if nums[start] <= target and target < nums[mid]:
                    end = mid - 1
                else:
                    start = mid + 1
            # right sorted
            else:
                if nums[mid] < target and target <= nums[end]:
                    start = mid + 1
                else:
                    end = mid - 1
        return False
